app/zapier/files/upload.js

```diff
@@ -0,0 +1,132 @@
+
+//
+// (c) Copyright Ascensio System SIA 2023
+//
+
+// @ts-check
+
+const contentDisposition = require("content-disposition")
+const FormData = require("form-data")
+
+/**
+ * @typedef {import("../../docspase/files/files.js").ChunkData} ChunkData
+ * @typedef {import("../../zapier/files/actions.js").UploadFileFields} UploadFileFields
+ */
+
+/**
+ * @typedef {Object} ErrorData
+ * @property {string} message
+ * @property {string} stack
+ */
+
+/**
+ * @typedef {Object} UploadFileData
+ * @property {number} id
+ * @property {number} folderId
+ * @property {number} version
+ * @property {string} title
+ * @property {boolean} uploaded
+ * @property {FileData} file
+ */
+
+/**
+ * @typedef {Object} Headers
+ * @property {string} fileName
+ * @property {number} fileSize
+ */
+
+/**
+ * @typedef {import("./actions.js").FileData} FileData
+ */
+
+class Upload {
+  /**
+   * @param {ZObject} z
+   */
+  constructor(z) {
+    this.z = z
+  }
+
+  /**
+   * @param {string} url
+   * @returns {Promise<string>}
+   */
+  async stash(url) {
+    let zapierUrl
+    if (url.startsWith("https://zapier.com/engine/hydrate/")) {
+      zapierUrl = url
+    } else {
+      const filePromise = this.z.request({ url, raw: true })
+      zapierUrl = await this.z.stashFile(filePromise)
+    }
+    return zapierUrl
+  }
+
+  /**
+   * @param {string} url
+   * @returns {Promise<Headers>}
+   */
+  async headers(url) {
+    const { headers } = await fetch(url, { method: "HEAD" })
+    const length = headers.get("content-length")
+    const disposition = headers.get("content-disposition")
+    const fileSize = length ? parseInt(length) : 0
+    const fileName = disposition ? contentDisposition.parse(disposition).parameters.filename : "File from Zapier"
+    return { fileName, fileSize }
+  }
+
+  /**
+   * @param {(arg0: ChunkData) => Promise<UploadFileData>} uploadChunk
+   * @param {string} uploadLocation
+   * @param {string} stashedFile
+   * @param {string} fileName
+   * @param {number} fileSize
+   * @returns {Promise<UploadFileData>}
+   */
+  async upload(uploadChunk, uploadLocation, stashedFile, fileName, fileSize) {

```

Comment:
I find myself confused by the abundance of parameters available. It would be greatly appreciated if you could find a way to simplify them, although this is not mandatory.

Reply:
> I find myself confused by the abundance of parameters available. It would be greatly appreciated if you could find a way to simplify them, although this is not mandatory.

I created a bodyUpload object with properties `fileName, fileSize, fileStash`